Update EIP-4972: Move status from draft to review (#5248)
ethereum/EIPs: EIPS/eip-4972.md

2. an address derived from the name to receive tokens; and
3. owner(s) of the name that can transfer the token.

The name should be human-readable and can be easily recognized socially. An example is an username of a name service or an email address of a centralized platform such as Gmail. The name-derived address (NDA) is a normal Ethereum address that should not collide with the existing addresses of EOA/CA. Since we cannot use NDA as msg.sender, the right to transfer the tokens of the NDA is controlled by the owner/owners of the name. The name to owner/owners mapping is managed by an on-chain name service, and the owner/owners should be EOA/CA, which can be the addresses of 3-rd custodians (e.g., Facebook or Google) or self-custodian. By changing the owner of the name to the EOA of the user (can be done by requesting the custodian), the NDA becomes self-custodian, and no one should be able to transfer the assets unless the approved by the self-custodian user. 

> ### Name Representation

Reviewer: ```suggestion
### Name Representation

```

A name is represented by a bytes array which is ABI encoded. It **MAY** contain metadata of the name such as the name service the name belongs to.  Example of the name is "vitalik.eth" or "qizhou@quarkchain.org".

> ### Interface

Reviewer: ```suggestion
### Interface

```

> #### INameOwnedAccount

Reviewer: ```suggestion
#### INameOwnedAccount

```

> ```

Reviewer: ```suggestion
```solidity
```

interface INameOwnedAccount {
    /// @notice This function resolves a name to NDA
    /// @dev The implementation SHOULD avoid collision between name 
    /// derived address and EOA/CA
    /// @dev Return address(0) if the name is not supported
    function addressOfName(bytes memory _name) public view returns(address);

    /// @notice This function returns true if and only if _nda represents a name
    /// and the _operator is the owner of the name
    /// @dev The ownership MAY be defined by a name service such as ENS
    function isNameOwner(address _nda, address _operator) public view returns(bool);
}
```

> #### `IERC721NOA`

Reviewer: ```suggestion
#### `IERC721NOA`

```

> ```

Reviewer: ```suggestion
```solidity
```

interface IERC721NOA is IERC721, INameOwnedAccount { }
```

The existing `transferFrom` and `safeBatchTransferFrom` functions MUST assume caller is approved to manage tokens of `_from` if `_from` represents a name and the caller is the owner of the name.

> ```

Reviewer: ```suggestion
```solidity
```

function transferFrom(address _from, address _to, uint256 _tokenId) external payable;
function safeTransferFrom(address _from, address _to, uint256 _tokenId, bytes data) external payable;
function safeTransferFrom(address _from, address _to, uint256 _tokenId) external payable;
```

The existing `isApprovedForAll` function of `IERC721` MUST return true if the `_owner` represents a name and `_operator` is the owner of the name.

> ```

Reviewer: ```suggestion
```solidity
```

function isApprovedForAll(address _owner, address _operator) external view returns (bool);
```

The existing `ownerOf` function at `IERC721` should return the NOA owning the token.

> ```

Reviewer: ```suggestion
```solidity
```

function ownerOf(uint256 _tokenId) external view returns (address);
```

#### `IERC1155NOA`

> ```

Reviewer: ```suggestion
```solidity
```

interface IERC1155NOA is IERC1155, INameOwnedAccount  {}
```

The existing `safeTransferFrom` and `safeBatchTransferFrom` functions MUST assume caller is approved to manage tokens of `_from` if `_from` represents a name and the caller is the owner of the name.

> ```

Reviewer: ```suggestion
```solidity
```

function safeTransferFrom(address _from, address _to, uint256 _id, uint256 _value, bytes calldata _data) external;
function safeBatchTransferFrom(address _from, address _to, uint256[] calldata _ids, uint256[] calldata _values, bytes calldata _data) external;
```

The existing `isApprovedForAll` function of `IERC1155` MUST return true if the `_owner` represents a name and `_operator` is the owner of the name.

> ```

Reviewer: ```suggestion
```solidity
```

function isApprovedForAll(address _owner, address _operator) external view returns (bool);
```

> #### `IERC20NOA`

Reviewer: ```suggestion
#### `IERC20NOA`

```

> ```

Reviewer: ```suggestion
```solidity
```

interface IERC20NOA is IERC20, INameOwnedAccount {}
```

The existing `transferFrom` function MUST assume caller is approved to manage tokens of `_from` if `_from` represents a name and the caller is the owner of the name.

> ```

Reviewer: ```suggestion
```solidity
```

function transferFrom(address _from, address _to, uint256 _value) public returns (bool success)
```

The existing `allowance ` function MUST return max value of uint256 if `_owner` represents a name and the `_spender` is the owner of the name.

> ```

Reviewer: ```suggestion
```solidity
```

function allowance(address _owner, address _spender) public view returns (uint256 remaining)
```

> ## Rationale

Reviewer: ```suggestion
## Rationale

```

The isNameOwner function is sufficient for authenticating the message sender. One can verify the owner by looking up the owner of an NDA from a name service.

The addressOfName interface decouples the implementation from specific NDA encoding algorithm. How to encode a name to an NDA is not defined by this standard, as long as the generated NDA does not collide with EOA/CA address space.

## Backwards Compatibility

> The new account type is compatible with existing ERC token standards.

Reviewer: ```suggestion

The new account type is compatible with existing ERC token standards.
```

> ## Reference Implementation

Reviewer: ```suggestion
## Reference Implementation

```

> ### Name Derived Address (INameOwnedAccount.addressOfName())

Reviewer: ```suggestion
### Name Derived Address (INameOwnedAccount.addressOfName())

```

> Each token contract can decide how to encode the name to an NDA as long as the address space of NDAs does not overlap with existing EOA/CA spaces. We can follow the similar rule of CREATE2 opcode to compute the NDA from a name service and the hash of a username as 

Reviewer: ```suggestion

Each token contract can decide how to encode the name to an NDA as long as the address space of NDAs does not overlap with existing EOA/CA spaces. We can follow the similar rule of CREATE2 opcode to compute the NDA from a name service and the hash of a username as 
```

> ```

Reviewer: ```suggestion
```solidity
```

> address(keccak256(0xff, keccak256(“eip-4972.addressOfName”), nameServiceAddress, keccak256(username))).

Reviewer: ```suggestion
address(keccak256(0xff, keccak256("eip-4972.addressOfName"), nameServiceAddress, keccak256(username))).
```

```

This can ensure it will not collide with existing smart contract account addresses.

### Ownership of a Name (INameOwnedAccount.isNameOwner())

> Given an NDA, we need to lookup the owner from a name service and compare it with the message sender. Note that current ENS only supports namehash (32 bytes) to owner translation, and thus we may need an additional helper contract to map an NDA to a namehash for ENS.

Reviewer: ```suggestion

Given an NDA, we need to look up the owner from a name service and compare it with the message sender. Note that currently, ENS only supports namehash (32 bytes) to owner translation, and thus we may need an additional helper contract to map an NDA to a namehash for ENS.
```

## Security Considerations

> No security considerations were found.

Reviewer: ```suggestion

No security considerations were found.
```

}
```

> #### IEIP-721NOA

Author: Thanks for the comment.  Update the PR accordingly.
